### PR TITLE
[MNT] improved changelog generator: group PRs by module labels like Forecasting and Classification

### PR DESCRIPTION
**Pull request by sinemkilicdere**
1. PR addresses #8128, which requests adding subsections to the changelog based on module labels like module:forecasting, module:classification.

2. Firstly, the changelog groups PRs into level categories like:

Enhancements
Fixes
Maintenance
Documentation

But all PRs go under the Other subsection, even if they have proper module: labels.

3.What issue am I facing?
Although the logic for generating module based subsections is in place, everything is still grouped under Other.

4. After running for tets:
python build_tools/changelog.py

The output shows Enhancements and other top level categorieys, but no subsections like “Forecasting” or “Classification” appear

assign_prs()
Modified to extract and return pr_module dictionary mapping each PR index to its module: label

render_changelog()
Defined LABEL_TO_SUBSECTION, which maps module labels to human readable section titles.

5.Fully test that module based subsections are rendered correctly

##### For all contributions
- [ ] I've added myself to the [list of contributors](https://github.com/sktime/sktime/blob/main/CONTRIBUTORS.md) with any new badges I've earned :-)
  How to: add yourself to the [all-contributors file](https://github.com/sktime/sktime/blob/main/.all-contributorsrc) in the `sktime` root directory (not the `CONTRIBUTORS.md`). Common badges: `code` - fixing a bug, or adding code logic. `doc` - writing or improving documentation or docstrings. `bug` - reporting or diagnosing a bug (get this plus `code` if you also fixed the bug in the PR).`maintenance` - CI, test framework, release.
  See here for [full badge reference](https://github.com/all-contributors/all-contributors/blob/master/docs/emoji-key.md)
- [ ] Optionally, for added estimators: I've added myself and possibly to the `maintainers` tag - do this if you want to become the owner or maintainer of an estimator you added.
  See here for further details on the [algorithm maintainer role](https://www.sktime.net/en/latest/get_involved/governance.html#algorithm-maintainers).
- [ ] The PR title starts with either [ENH], [MNT], [DOC], or [BUG]. [BUG] - bugfix, [MNT] - CI, test framework, [ENH] - adding or improving code, [DOC] - writing or improving documentation or docstrings.

##### For new estimators
- [ ] I've added the estimator to the API reference - in `docs/source/api_reference/taskname.rst`, follow the pattern.
- [ ] I've added one or more illustrative usage examples to the docstring, in a pydocstyle compliant `Examples` section.
- [ ] If the estimator relies on a soft dependency, I've set the `python_dependencies` tag and ensured
  dependency isolation, see the [estimator dependencies guide](https://www.sktime.net/en/latest/developer_guide/dependencies.html#adding-a-soft-dependency).

### CODE REVIEW

**Review by fkiraly — CHANGES_REQUESTED** [review]
Nice!

I suggest you also set up code formatting locally, here is the guide.

**Comment by jgyasu** [comment]
Guide for code formatting: [https://www.sktime.net/en/latest/developer_guide/coding_standards.html](https://www.sktime.net/en/latest/developer_guide/coding_standards.html)

**Comment by fkiraly** [comment]
I think this did not work - can you try install and run `pre-commit`?

**Comment by sinemkilicdere** [comment]
> I think this did not work - can you try install and run `pre-commit`?

I have already installed and ran before last push. and it all passed. what could be problem?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Really great, this is very useful!

Here is the current printout, I add some comments based on that.

```
Latest release v0.38.4 was published at 2025-07-17 16:23:37+00:00
Found 10 merged PRs since last release

Fixes
~~~~~


Base Framework
^^^^^^^^^^^^^^

* [BUG] ensure ``deps`` dependency inspection is compatible with disjunctions in estimator dependency tags (:pr:`8585`) :user:`fkiraly`

Clustering
^^^^^^^^^^

* [BUG] decouple classifier, clusterer, regressor extension contract from ``n_jobs`` parameter (:pr:`8549`) :user:`fkiraly`

Time Series Classification
^^^^^^^^^^^^^^^^^^^^^^^^^^

* [BUG] decouple classifier, clusterer, regressor extension contract from ``n_jobs`` parameter (:pr:`8549`) :user:`fkiraly`

Regression
^^^^^^^^^^

* [BUG] decouple classifier, clusterer, regressor extension contract from ``n_jobs`` parameter (:pr:`8549`) :user:`fkiraly`

Maintenance
~~~~~~~~~~~


Other
^^^^^

* Revert "insta-release for 0.38.4" (:pr:`8573`) :user:`fkiraly`
* [MNT] separate dependency set for notebooks, reduce dependencies in notebook CI (:pr:`8598`) :user:`fkiraly`

Enhancements
~~~~~~~~~~~~


Forecasting
^^^^^^^^^^^

* [ENH] ``LTSFNLinearForecaster``,`` LTSFDLinearForecaster``,  and ``LTSFLinearForecaster`` ``get_test_params`` returns two sets (:pr:`8528`) :user:`thearshkumar`
* [ENH] Added tag to PytorchForecastingTFT and removed from tests/_config (:pr:`8559`) :user:`thearshkumar`

Plotting & Utilities
^^^^^^^^^^^^^^^^^^^^

* [ENH] add more common packages in standard ``show_versions`` return (:pr:`8484`) :user:`fkiraly`

Documentation
~~~~~~~~~~~~~


Other
^^^^^

* &fkiraly [DOC] Add guide for module-level soft dependency imports (#7690) (:pr:`8386`) :user:`aarsh1a`
* [DOC] fix import location of ``_safe_import`` in developer guide (:pr:`8577`) :user:`fkiraly`

Base Framework
^^^^^^^^^^^^^^

* [DOC] improve formatting and clarity in ``python_dependencies`` tag description (:pr:`8517`) :user:`fkiraly`

Contributors
~~~~~~~~~~~~

:user:`aarsh1a`,
:user:`fkiraly`,
:user:`thearshkumar`
```

* could you kindly ensure the big sections Enhancements, Fixes, Documentation, Maintenance always appear in the same order as for previous changelogs?
* I think there is one too many newline after the section header
* Documentation and Maintenance should be flat - sub-headers only for Enhancements and Fixes.

**Comment by fkiraly** [comment]
Thanks! Please look at past changelogs though, enhancements should be first.

As a reference have a look at `changelog_header_template.rst`

**Comment by sinemkilicdere** [comment]
> Thanks! Please look at past changelogs though, enhancements should be first.
> 
> As a reference have a look at `changelog_header_template.rst`

I have fixed all requirements based on feedback. thanks for review

(sktime-dev-env) C:\Users\CIS440\sktime>python build_tools/changelog.py
Latest release v0.38.4 was published at 2025-07-17 16:23:37+00:00
Found 16 merged PRs since last release

Enhancements
~~~~~~~~~~~~

Other
^^^^^
* [ENH] Jarque-Bera, variance break, and Ljung-Box tests/statistics (:pr:`8017`) :user:`HarshvirSandhu`

Base Framework
^^^^^^^^^^^^^^
* [BUG] fix ``ray`` backend unintended side effects on ``backend_params`` by replacing ``list.pop`` with ``list.get`` (:pr:`8162`) :user:`LHoelper`

Time Series Classification
^^^^^^^^^^^^^^^^^^^^^^^^^^
* [ENH] ensure that any change in ``classification.sklearn`` triggers the classification tests (:pr:`8609`) :user:`fkiraly`

Forecasting
^^^^^^^^^^^
* [ENH] ``LTSFNLinearForecaster``,`` LTSFDLinearForecaster``,  and ``LTSFLinearForecaster`` ``get_test_params`` returns two sets (:pr:`8528`) :user:`thearshkumar`
* [ENH] Added tag to PytorchForecastingTFT and removed from tests/_config (:pr:`8559`) :user:`thearshkumar`
* [ENH] ``darts`` compatibility patch for 0.35.0 and higher (:pr:`8584`) :user:`fkiraly`

Plotting & Utilities
^^^^^^^^^^^^^^^^^^^^
* [ENH] add more common packages in standard ``show_versions`` return (:pr:`8484`) :user:`fkiraly`

Fixes
~~~~~

Base Framework
^^^^^^^^^^^^^^
* [BUG] ensure ``deps`` dependency inspection is compatible with disjunctions in estimator dependency tags (:pr:`8585`) :user:`fkiraly`

Clustering
^^^^^^^^^^
* [BUG] decouple classifier, clusterer, regressor extension contract from ``n_jobs`` parameter (:pr:`8549`) :user:`fkiraly`

Time Series Classification
^^^^^^^^^^^^^^^^^^^^^^^^^^
* [BUG] decouple classifier, clusterer, regressor extension contract from ``n_jobs`` parameter (:pr:`8549`) :user:`fkiraly`

Regression
^^^^^^^^^^
* [BUG] decouple classifier, clusterer, regressor extension contract from ``n_jobs`` parameter (:pr:`8549`) :user:`fkiraly`

Documentation
~~~~~~~~~~~~~
* [DOC] fix import location of ``_safe_import`` in developer guide (:pr:`8577`) :user:`fkiraly`
* &fkiraly [DOC] Add guide for module-level soft dependency imports (#7690) (:pr:`8386`) :user:`aarsh1a`
* [DOC] improve formatting and clarity in ``python_dependencies`` tag description (:pr:`8517`) :user:`fkiraly`

Maintenance
~~~~~~~~~~~
* [MNT] update notebook CI python version to 3.11 (:pr:`8604`) :user:`fkiraly`
* [MNT] separate dependency set for notebooks, reduce dependencies in notebook CI (:pr:`8598`) :user:`fkiraly`
* Revert "insta-release for 0.38.4" (:pr:`8573`) :user:`fkiraly`

Contributors
~~~~~~~~~~~~

:user:`aarsh1a`,
:user:`fkiraly`,
:user:`HarshvirSandhu`,
:user:`julian-fong`,
:user:`LHoelper`,
:user:`thearshkumar`

(sktime-dev-env) C:\Users\CIS440\sktime>

File: .all-contributorsrc [hunk 1]
```diff
@@ -41,6 +41,20 @@
         "video"
       ]
     },
+    {
```
Comment (fkiraly) [inline .all-contributorsrc]: minor comments about the contributions file:
* please move your record to the end. The entries are roughly chronological.
* `code` is fine, thanks for your contribution!
* `maintenance` is fine since this concerns the CI/release
* `doc` cannot be added with PR. You can add it with the other PR related to docstrings.
* `test` is not part of this PR.
* `ideas` is not part of this PR
* `question` is not justified afaik since you have not answered user or contributor questions on the community channels (yet)

File: docs/source/examples [hunk 2]
```diff
@@ -1 +1 @@
-../../examples
```
Comment (fkiraly) [inline docs/source/examples]: can you please revert your change to this file?

File: build_tools/changelog.py [hunk 3]
```diff
@@ -184,7 +238,7 @@ def render_changelog(prs, assigned):
 
     pulls = fetch_pull_requests_since_last_release()
     print(f"Found {len(pulls)} merged PRs since last release")
-    assigned = assign_prs(pulls, categories)
+    assigned, pr_module = assign_prs(pulls, categories)
```
Comment (fkiraly) [inline build_tools/changelog.py]: why does this additional return exist, if it is never used?

Comment (sinemkilicdere) [reply build_tools/changelog.py]: it was for testing func. removed and fixed regarding comments. tnx

**Review by fkiraly — CHANGES_REQUESTED** [review]
Great!

Comments not addressed:

* the return `pr_module` seems unnecessary and should be removed.
* too many newlines after section headers
* subsection headers should be in same order as `changelog_header_template.rst`

I also recommend you try to run this locally, and see how this prints on the main branch - this will also allow you to discover the issues yourself.

**Comment by sinemkilicdere** [comment]
Locally, my branch had many unrelated commits. I cleaned up the history to avoid pushing unrelated changes every time and made the final updates before this last push. thats whay there is only one commit

File: build_tools/changelog.py [hunk 4]
```diff
@@ -159,27 +159,94 @@ def render_row(pr):
...
     """Render changelog."""
+    # sourcery skip: use-named-expression
+    LABEL_TO_SUBSECTION = {
+        "module:forecasting": "Forecasting",
+        "module:classification": "Time Series Classification",
+        "module:regression": "Regression",
+        "module:transformations": "Transformations",
+        "module:datatypes": "Datatypes",
+        "module:alignment": "Alignment",
+        "module:annotation": "Annotation",
```
Comment (fkiraly) [inline build_tools/changelog.py]: the name is deprecated, we should use `module:detection` and `Detection` instead

Plus, please have a look at the template, it should be `Time Series Clustering`, etc, you can infer the correct names of the headers from there.

File: build_tools/changelog.py [hunk 5]
```diff
@@ -159,27 +159,94 @@ def render_row(pr):
...
+        "module:proba": "Probability Estimation",
+        "module:clustering": "Clustering",
+        "module:plotting&utilities": "Plotting & Utilities",
+        "module:base-framework": "Base Framework",
+    }
+    # more can be added also
+    # labels start with module
+    # Subsection title comes after label
+
+    MODULE_ORDER = [
```
Comment (fkiraly) [inline build_tools/changelog.py]: The set of names in this list is not the same as the set of names above.

Comment (fkiraly) [reply build_tools/changelog.py]: it still is not

File: build_tools/changelog.py [hunk 6]
```diff
@@ -159,27 +159,91 @@ def render_row(pr):
...
+
+        # Group PRs by module labels
+        if title in ["Enhancements", "Fixes"]:
+            subsection_map = defaultdict(list)
+            for pr in pr_group:
+                labels = [label["name"] for label in pr["labels"]]
+                added = False
+
+                for label in labels:
+                    # print("hello",label)
```
Comment (fkiraly) [inline build_tools/changelog.py]: remove commented out lines

File: build_tools/changelog.py [hunk 7]
```diff
@@ -159,27 +159,91 @@ def render_row(pr):
...
+            continue
+        print(title)
+        print("~" * len(title))
+        print()
+
+        # Group PRs by module labels
+        if title in ["Enhancements", "Fixes"]:
+            subsection_map = defaultdict(list)
+            for pr in pr_group:
+                labels = [label["name"] for label in pr["labels"]]
```
Comment (fkiraly) [inline build_tools/changelog.py:236]: may I suggest ot structure this better by movin this out into functions?

**Review by fkiraly — CHANGES_REQUESTED** [review]
Really great!

I left some small comments above:

* the tag-to-header dict is incomplete
* the header order list does not contain all headers
* request to structure the printout code better by moving parts into functions

**Comment by sinemkilicdere** [comment]
> Thanks! Looks like almost done!
> 
> 
> 
> Two change requests:
> 
> 
> 
> * the labels in the dictionary `LABELS_TO_SUBSECTION` do not match the actual labels. Let me know if you do not know how to get the GH labels (e.g., if you have been using AI to guess them instead of looking them up)
> 
> * The subsection headers do not have enough newlines, please compare with prior changelogs. They are all printed without empty line before/after now.
> 
> 
> 
> I would also advise that you try run the utility and compare with actual changelogs from the past, in `changelog.rst`.

I wrote all from the source of changelog_header_template.rst file. I guess for that reference you just meant for module_labels. I have used both, any suggestion I can take a look the list

**Comment by fkiraly** [comment]
to see all labels, click "labels" in the GitHub issues or PR UI:
image

This contains more labels than are used in the changelog, but it should show you how they are exactly spelled out.

You can also get them as a list of str by using the GitHub API, as in the changelog generator script (before the matching/translation)

**Comment by sinemkilicdere** [comment]
1. last changes have been made. I used --amend under same commit. labels are taken from https://github.com/sktime/sktime/labels and match with modules - from changelog_header_template.rst .. if no match, I categorized under `Other`. 
2. Spacing issue has been solved regarding changelog.rst.
Thanks for review

**Review by fkiraly — CHANGES_REQUESTED** [review]
great, thanks!

1. I think the label issue is resolved, excellent!
2. I do not think the spacing issue is entirely resolved - try running it on the current set of PR

**Review by fkiraly — DISMISSED** [review]
Thanks!

I merged this with #8132 to also give @Mujeeb4 credit for the effort.

No major changes, except some additional features from @Mujeeb4´s PR are added (e.g., dependabot handling)
